src/pipe/modules/lenswarp/main.c

```diff
@@ -0,0 +1,151 @@
+#include "modules/api.h"
+#include "core/log.h"
+#include "config.h"
+
+static void generate_radius_lut_dng(
+    dt_image_params_t *img_param,
+    float             *lut)
+{
+  dt_lens_corr_dng_t *dng = &img_param->lens_corr.params.dng;
+
+  if(dng->warp_rect_coef[0].kt0 != 0 || dng->warp_rect_coef[1].kt1 != 0
+    || dng->warp_rect_coef[1].kt0 != 0 || dng->warp_rect_coef[1].kt1 != 0
+    || dng->warp_rect_coef[2].kt0 != 0 || dng->warp_rect_coef[2].kt1 != 0)
+  {
+    dt_log(s_log_err, "[lenswarp] Ignoring WarpRectilinear tangential coefficients");
+  }
+
+  // For WarpRectilinear, radius 1 is defined as the distance from the center to
+  // the furthest corner of the ActiveArea. For the LUT we want to define radius
+  // 1 as the corner of the crop area.
+
+  // pixel location of center in the uncropped ActiveArea
+  float cx = dng->warp_rect_center_x * (dng->active_area[3] - dng->active_area[1]);
+  float cy = dng->warp_rect_center_y * (dng->active_area[2] - dng->active_area[0]);
+  // distance to furthest corner of uncropped ActiveArea
+  float dng_dist = hypotf(
+    MAX(cx, (dng->active_area[3] - dng->active_area[1]) - cx),
+    MAX(cy, (dng->active_area[2] - dng->active_area[0]) - cy));
+  // distance to furthest corner of cropped image
+  float crop_dist = hypotf(
+    MAX(cx - img_param->crop_aabb[0], img_param->crop_aabb[2] - cx),
+    MAX(cy - img_param->crop_aabb[1], img_param->crop_aabb[3] - cy));
+  float r_scale = crop_dist / dng_dist;
+
+  for(int ch = 0; ch < 3; ch++)
+  {
+    dt_warp_rect_coef_t *c = &dng->warp_rect_coef[ch];
+    for(int i = 0; i < LUT_SIZE; i++)
+    {
+      float r = (float)i / (LUT_SIZE - 1); // radius in the undistorted output image
+      float rs2 = r * r * r_scale;
+      lut[i * 4 + ch] = c->kr0 + rs2 * (c->kr1 + rs2 * (c->kr2 + rs2 * c->kr3));
```

Comment:
yeah see general comment. this is not nearly enough to make a gpu sweat per pixel. also since it appears only even powers of r are needed, maybe the square root above (in the hypot) is not actually required to be computed at all, so the special function unit would be idle too.